### ISSUE TITLE
Missing REDIS support leads to inaccessibility of the server

### ISSUE THREAD

**Issue by loma-one** [friendica/friendica#13318]
- [x] I have searched open and closed issues for duplicates

### Bug Description

Missing REDIS support leads to inaccessibility of the server (error 500)

### Steps to Reproduce

```
2023-07-29T04:51:20Z app [ERROR]: Uncaught Exception Friendica\Core\Hooks\Exceptions\HookInstanceException: "The class with the name Redis isn't registered for the class or interface Friendica\Core\Cache\Capability\ICanCache" at /var/www/html/friend/src/Core/Hooks/Model/DiceInstanceManager.php line 65 {"exception":"Friendica\\Core\\Hooks\\Exceptions\\HookInstanceException: The class with the name Redis isn't registered for the class or interface Friendica\\Core\\Cache\\Capability\\ICanCache in /var/www/html/friend/src/Core/Hooks/Model/DiceInstanceManager.php:65
Stack trace:
#0 /var/www/html/friend/src/Core/Cache/Factory/Cache.php(100): Friendica\\Core\\Hooks\\Model\\DiceInstanceManager->create()
#1 /var/www/html/friend/src/Core/Cache/Factory/Cache.php(84): Friendica\\Core\\Cache\\Factory\\Cache->create()
#2 /var/www/html/friend/vendor/level-2/dice/Dice.php(150): Friendica\\Core\\Cache\\Factory\\Cache->createLocal()
#3 /var/www/html/friend/vendor/level-2/dice/Dice.php(96): Dice\\Dice->Dice\\{closure}()
#4 /var/www/html/friend/vendor/level-2/dice/Dice.php(247): Dice\\Dice->create()
#5 /var/www/html/friend/vendor/level-2/dice/Dice.php(132): Dice\\Dice->Dice\\{closure}()
#6 /var/www/html/friend/vendor/level-2/dice/Dice.php(96): Dice\\Dice->Dice\\{closure}()
#7 /var/www/html/friend/index.php(47): Dice\\Dice->create()
#8 {main}"} - {"file":null,"line":null,"function":null,"request-id":"64c49ac81d71d","uid":"8b1076","process_id":262766}
```

Deactivating REDIS support restores the availability of the server

### Platform Info

Friendica Version: dev

Friendica Source: git

PHP version: 8.0

SQL version: 10.5

**Comment by loma-one** [comment]
From today after the update

```
2023-07-30T08:50:34Z app [ERROR]: Uncaught Exception Friendica\Core\Cache\Exception\CachePersistenceException: "Redis connection fails unexpectedly" at /var/www/html/friend/src/Core/Cache/Type/RedisCache.php line 78 {"exception":"RedisException: Connection refused in /var/www/html/friend/src/Core/Cache/Type/RedisCache.php:66
Stack trace:
#0 /var/www/html/friend/src/Core/Cache/Type/RedisCache.php(66): Redis->connect()
#1 [internal function]: Friendica\\Core\\Cache\\Type\\RedisCache->__construct()
#2 /var/www/html/friend/vendor/level-2/dice/Dice.php(132): ReflectionMethod->invokeArgs()
#3 /var/www/html/friend/vendor/level-2/dice/Dice.php(96): Dice\\Dice->Dice\\{closure}()
#4 /var/www/html/friend/src/Core/Hooks/Model/DiceInstanceManager.php(68): Dice\\Dice->create()
#5 /var/www/html/friend/src/Core/Cache/Factory/Cache.php(100): Friendica\\Core\\Hooks\\Model\\DiceInstanceManager->create()
#6 /var/www/html/friend/src/Core/Cache/Factory/Cache.php(84): Friendica\\Core\\Cache\\Factory\\Cache->create()
#7 /var/www/html/friend/vendor/level-2/dice/Dice.php(150): Friendica\\Core\\Cache\\Factory\\Cache->createLocal()
#8 /var/www/html/friend/vendor/level-2/dice/Dice.php(96): Dice\\Dice->Dice\\{closure}()
#9 /var/www/html/friend/vendor/level-2/dice/Dice.php(247): Dice\\Dice->create()
#10 /var/www/html/friend/vendor/level-2/dice/Dice.php(132): Dice\\Dice->Dice\\{closure}()
#11 /var/www/html/friend/vendor/level-2/dice/Dice.php(96): Dice\\Dice->Dice\\{closure}()
#12 /var/www/html/friend/index.php(47): Dice\\Dice->create()
#13 {main}

Next Friendica\\Core\\Cache\\Exception\\CachePersistenceException: Redis connection fails unexpectedly in /var/www/html/friend/src/Core/Cache/Type/RedisCache.php:78
Stack trace:
#0 [internal function]: Friendica\\Core\\Cache\\Type\\RedisCache->__construct()
#1 /var/www/html/friend/vendor/level-2/dice/Dice.php(132): ReflectionMethod->invokeArgs()
#2 /var/www/html/friend/vendor/level-2/dice/Dice.php(96): Dice\\Dice->Dice\\{closure}()
#3 /var/www/html/friend/src/Core/Hooks/Model/DiceInstanceManager.php(68): Dice\\Dice->create()
#4 /var/www/html/friend/src/Core/Cache/Factory/Cache.php(100): Friendica\\Core\\Hooks\\Model\\DiceInstanceManager->create()
#5 /var/www/html/friend/src/Core/Cache/Factory/Cache.php(84): Friendica\\Core\\Cache\\Factory\\Cache->create()
#6 /var/www/html/friend/vendor/level-2/dice/Dice.php(150): Friendica\\Core\\Cache\\Factory\\Cache->createLocal()
#7 /var/www/html/friend/vendor/level-2/dice/Dice.php(96): Dice\\Dice->Dice\\{closure}()
#8 /var/www/html/friend/vendor/level-2/dice/Dice.php(247): Dice\\Dice->create()
#9 /var/www/html/friend/vendor/level-2/dice/Dice.php(132): Dice\\Dice->Dice\\{closure}()
#10 /var/www/html/friend/vendor/level-2/dice/Dice.php(96): Dice\\Dice->Dice\\{closure}()
#11 /var/www/html/friend/index.php(47): Dice\\Dice->create()
#12 {main}"} - {"file":null,"line":null,"function":null,"request-id":"64c6245acaab8","uid":"fb0738","process_id":403958}
```

**Comment by nupplaphil** [comment]
Connection refused o_O

Do you have any special settings like a specific db, a password or something like that?

**Comment by AlfredSK** [comment]
Hmm. Seems to work here. I have configured a redis db number and I'm accessing redis via Unix socket, not IP and port. No password used.
I'll check my log too.

Edit: Nope. No errors in application log and PHP log.

**Comment by loma-one** [comment]
No special settings
```
'cache_driver' => 'Redis',
'redis_host' => '127.0.0.1',
'redis_port' => 0,
'redis_db' => 2,
'redis_password' => null,
```

**Comment by AlfredSK** [comment]
Have you tried to delete or comment out the password line?

**Comment by loma-one** [comment]
Deleting/commenting out the config setting for the password does not change the message. As soon as I activate Redis in the config, the server runs into a 500 error. The instance is then no longer accessible. Currently, I have deleted all Redis entries in order to access the instance.

Before the code change above, this was not a problem and other instances continue to run with these config entries. So the question is, what was changed in the code that now needs different settings and what are they?

**Comment by nupplaphil** [comment]
> No special settings
> ```
> 'cache_driver' => 'Redis',
> 'redis_host' => '127.0.0.1',
> 'redis_port' => 0,
> 'redis_db' => 2,
> 'redis_password' => null,
> ```

I take this settings and try it again

**Comment by AlfredSK** [comment]
Ok. I wonder about the port 0. Shouldn't that be the actual port Redis is listening on for connections?

**Comment by loma-one** [comment]
This is the port I have set up. 
```
                _._                                                  
           _.-``__ ''-._                                             
      _.-``    `.  `_.  ''-._           Redis 6.0.16 (00000000/0) 64 bit
  .-`` .-```.  ```\/    _.,_ ''-._                                   
 (    '      ,       .-`  | `,    )     Running in standalone mode
 |`-._`-...-` __...-.``-._|'` _.-'|     Port: 0
 |    `-._   `._    /     _.-'    |     PID: 599
  `-._    `-._  `-./  _.-'    _.-'                                   
 |`-._`-._    `-.__.-'    _.-'_.-'|                                  
 |    `-._`-._        _.-'_.-'    |           http://redis.io        
  `-._    `-._`-.__.-'_.-'    _.-'                                   
 |`-._`-._    `-.__.-'    _.-'_.-'|                                  
 |    `-._`-._        _.-'_.-'    |                                  
  `-._    `-._`-.__.-'_.-'    _.-'                                   
      `-._    `-.__.-'    _.-'                                       
          `-._        _.-'                                           
              `-.__.-'                  
```

**Comment by loma-one** [comment]
I now use a different port instead of 0. It works. 
Since the port works without problems until stable, I have to make sure to change the Redis port before updating the main instance.

**Comment by MrPetovan** [comment]
Maybe we can add support for port=0 as well.

**Comment by nupplaphil** [comment]
> Maybe we can add support for port=0 as well.

I'd add 0 would be like "not set"..

**Comment by MrPetovan** [comment]
That's what I expect we do right now but apparently it needs to be a "set" value.

**Comment by AlfredSK** [comment]
When the port=0 is a set value what happens when I use Redis connection via UNIX socket and port=0 because the socket needs no port?
Should I remove the port line from the config file?

**Comment by MrPetovan** [comment]
This code prevents the use of `port = 0`.
https://github.com/friendica/friendica/blob/13cb8bb3955576ec53d79b1e204bb6cf41dddbd7/src/Core/Cache/Type/RedisCache.php#L64-L68
It isn't a big deal to change.

**Comment by MrPetovan** [comment]
> When the port=0 is a set value what happens when I use Redis connection via UNIX socket and port=0 because the socket needs no port? Should I remove the port line from the config file?

Yes, it would use the default port value defined in `static/defaults.config.php` of `6379` and probably ignore it.

**Comment by MrPetovan** [comment]
Correction:
> _host_: string. can be a host, or the path to a unix domain socket.
> _port_: int (default is 6379, **should be -1 for unix domain socket**)
> _connectTimeout_: float, value in seconds (default is 0 meaning unlimited)
> _retryInterval_: int, value in milliseconds (optional)
> _readTimeout_: float, value in seconds (default is 0 meaning unlimited)
> _persistent_: mixed, if value is string then it used as persistend id, else value casts to boolean
> _auth_: mixed, authentication information
> _ssl_: array, SSL context options

From https://github.com/phpredis/phpredis#class-redis

**Comment by AlfredSK** [comment]
> should be -1 for unix domain socket

Aaaaah! Thank you! :-)

**Comment by nupplaphil** [comment]
Good question.. we should stick to the official redis nomenclature..

**Comment by Ramoonus** [comment]
Also happens in 2024.09
